src/Analyzers/CSharp/Analyzers/UseCollectionExpression/UseCollectionExpressionHelpers.cs

```diff
@@ -96,7 +96,12 @@ public static bool CanReplaceWithCollectionExpression(
 
         var parent = topMostExpression.GetRequiredParent();
 
-        if (!IsInTargetTypedLocation(semanticModel, topMostExpression, cancellationToken))
+        var targetType = topMostExpression.GetTargetType(semanticModel, cancellationToken);
+        if (targetType is null or IErrorTypeSymbol)
+            return false;
+
+        // (X[])[1, 2, 3] is target typed.  `(X)[1, 2, 3]` is currently not (because it looks like indexing into an expr).
+        if (topMostExpression.Parent is CastExpressionSyntax castExpression && castExpression.Type is IdentifierNameSyntax)
```

Comment:
pulled out of the core helper (which i moved).  it is a "use collection expr" specific test.